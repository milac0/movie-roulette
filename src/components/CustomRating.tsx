import React, { useContext, useEffect } from "react";
import axios from "axios";
import { css } from "emotion";
import Rating from "@material-ui/lab/Rating";
import { isAuthenticated } from "./../helpers/index";
import { useState } from "react";
import { UserContext } from "../context/UserContext";

const msg = css`
  color: #333333;
  font-size: 0.75rem;
  font-weight: 500;
  margin: 0;
  margin-left: 0.5em;
  max-width: 400px;
`;

const rateElement = css`
  & span {
    font-size: 40px;
  }
`;

interface Props {
  movieid: string | undefined;
}

const CustomRating: React.FC<Props> = ({ movieid }) => {
  const { user } = useContext(UserContext);
  const [message, setMessage] = useState();
  const [rate, setRate] = useState<number | null>(0);
  useEffect(() => {
    (async () => {
      // no other way to fetch guest users ratings
      const ratedMovies = (
        await axios.get(
          `/guest_session/${user.guest_session_id}/rated/movies?api_key=${process.env.API_KEY}&language=en-US&sort_by=created_at.asc`
        )
      ).data.results;
      if (ratedMovies) {
        const movie = ratedMovies.filter(
          rated => rated.id.toString() === movieid
        );
        movie.length > 0 ? setRate(movie[0].rating) : setRate(null);
      }
    })();
  }, []);
  const handleRating = async (value: number | null) => {
    if (isAuthenticated(user.expires_at)) {
      try {
        const response = (
          await axios.post(
            `/movie/${movieid}/rating?api_key=${process.env.API_KEY}&guest_session_id=${user.guest_session_id}`,
            { value }
          )
        ).data;
        setRate(value);
        setMessage(response.status_message);
      } catch (err) {
        setRate(null);
        setMessage("Error while submiting rating. Try again later.");
      }
    } else {
      setRate(null);
      setMessage("You do not have permissions to access the service. Log in.");
    }
  };

  return (
    <div>
      <Rating
        className={rateElement}
        name="movie-stars"
        max={10}
        value={rate}
        onChange={(e, value) => handleRating(value)}
        size="large"
      />
      <h2 className={msg}>{message || null}</h2>
    </div>
  );
};

export default CustomRating;
