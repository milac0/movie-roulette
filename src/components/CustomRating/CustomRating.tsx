import React, { useContext, useEffect } from "react";
import axios from "axios";
import Rating from "@material-ui/lab/Rating";
import { useState } from "react";
import { UserContext } from "../../context/UserContext";
import styles from "./customRating.scss";

interface Props {
  movieid: string | undefined;
}

const CustomRating: React.FC<Props> = ({ movieid }) => {
  const { user, isAuth } = useContext(UserContext);
  const [message, setMessage] = useState();
  const [rate, setRate] = useState<number | null>(0);
  useEffect(() => {
    (async () => {
      // no other way to fetch guest users ratings
      if (isAuth()) {
        try {
          const ratedMovies = (
            await axios.get(
              `/guest_session/${user.guest_session_id}/rated/movies?api_key=${process.env.API_KEY}&language=en-US&sort_by=created_at.asc`
            )
          ).data.results;
          if (ratedMovies) {
            const movie = ratedMovies.filter(
              rated => rated.id.toString() === movieid
            );
            movie.length > 0 ? setRate(movie[0].rating) : setRate(null);
          }
        } catch (err) {
          console.log(err);
        }
      }
    })();
  }, []);
  const handleRating = async (value: number | null) => {
    if (isAuth()) {
      try {
        await axios.post(
          `/movie/${movieid}/rating?api_key=${process.env.API_KEY}&guest_session_id=${user.guest_session_id}`,
          { value }
        );
        setRate(value);
        setMessage(null);
      } catch (err) {
        setRate(null);
        setMessage("Error while submiting rating. Try again later.");
      }
    } else {
      setRate(null);
      setMessage("You do not have permissions to access the service. Log in.");
    }
  };

  return (
    <div>
      <Rating
        className={styles.rateElement}
        name="movie-stars"
        max={10}
        value={rate}
        onChange={(e, value) => handleRating(value)}
        size="large"
      />
      <h2 className={styles.msg}>{message || null}</h2>
    </div>
  );
};

export default CustomRating;
