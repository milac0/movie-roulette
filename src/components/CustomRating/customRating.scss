.msg {
  color: #333333;
  font-size: 0.75rem;
  font-weight: 500;
  margin: 0;
  margin-left: 0.5em;
  max-width: 400px;
}

span {
  font-size: 1rem;
}

@media (min-width: 280px) {
  span {
    font-size: 1.5rem;
  }
}

@media (min-width: 370px) {
  span {
    font-size: 2rem;
  }
}

@media (min-width: 414px) {
  span {
    font-size: 2rem;
  }
}

@media (min-width: 614px) {
  span {
    font-size: 2.5rem;
  }
}
