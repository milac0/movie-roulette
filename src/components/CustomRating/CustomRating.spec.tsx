import React from 'react';
import renderer from 'react-test-renderer';

import CustomRating from './CustomRating';

describe('CustomRating renders correctly', () => {
    test('snapshot renders', () => {
        const props = {
            movieid: 419704
        }
        const tree = renderer.create(<CustomRating {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});