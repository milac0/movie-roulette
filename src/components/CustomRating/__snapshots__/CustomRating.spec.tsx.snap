// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomRating renders correctly snapshot renders 1`] = `
<div>
  <span
    aria-label={null}
    className="MuiRating-root"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    role={null}
  >
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-1"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        1 Star
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-1"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-2"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        2 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-2"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-3"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        3 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-3"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-4"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        4 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-4"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-5"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        5 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-5"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-6"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        6 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-6"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={6}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-7"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        7 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-7"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={7}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-8"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        8 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-8"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={8}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-9"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        9 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-9"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={9}
    />
    <label
      className="MuiRating-label"
      htmlFor="movie-stars-10"
    >
      <span
        className="MuiRating-icon MuiRating-iconEmpty"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      <span
        className="MuiRating-visuallyhidden"
      >
        10 Stars
      </span>
    </label>
    <input
      checked={false}
      className="MuiRating-visuallyhidden"
      id="movie-stars-10"
      name="movie-stars"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="radio"
      value={10}
    />
  </span>
  <h2
    className="msg"
  />
</div>
`;
